python/src/linalg.cpp

```diff
@@ -0,0 +1,218 @@
+
+// Copyright © 2023 Apple Inc.
+
+#include <limits>
+#include <numeric>
+#include <ostream>
+#include <string>
+#include <variant>
+
+#include <pybind11/iostream.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "mlx/linalg.h"
+#include "mlx/ops.h"
+#include "mlx/utils.h"
+
+#include "python/src/load.h"
+#include "python/src/overloaded.h"
+#include "python/src/utils.h"
+
+namespace py = pybind11;
+using namespace py::literals;
+
+using namespace mlx::core;
+using namespace mlx::core::linalg;
+
+void init_linalg(py::module_& parent_module) {
+  auto m =
+      parent_module.def_submodule("linalg", "mlx.core.linalg: Linear Algebra.");
+
+  m.def(
+      "norm",
+      [](const array& a,
+         const std::variant<std::monostate, int, double, std::string>& ord,
+         const std::variant<std::monostate, int, std::vector<int>>& axis,
+         const bool keepdims,
+         const StreamOrDevice stream) {
+        return std::visit(
+            overloaded{
+                [&](const double p) {
+                  if (std::isinf((float)p) || std::isinf(p)) {
+                    if (p > 0) {
+                      return norm(
+                          a,
+                          "inf",
+                          std::holds_alternative<std::monostate>(axis)
+                              ? std::vector<int>()
+                              : get_reduce_axes(axis, a.ndim()),
+                          keepdims,
+                          stream);
+                    }
+                    return norm(
+                        a,
+                        "-inf",
+                        get_reduce_axes(axis, a.ndim()),
+                        keepdims,
+                        stream);
+                  }
```

Comment:
Also this case handling should be in the C++ `norm`. As in it should support `std::numeric_limits<double>::infinity` (and negative inf). Then you don't need to do these cases for python and we get C++ support.

Reply:
Ideally, we can remove this case handling and use std::numeric_limits<double>::infinity in C++ code. However, what happens with std::numeric_limits<float>::infinity or  std::numeric_limits<bfloat16>::infinity? I think we should handle or 'infinities' in the same way. This is why I went for the string approach. If you think we should handle only std::numeric_limits<double>::infinity, we can move this to C++ and remove from bindings.